trl/trainer/rloo_trainer.py

```diff
@@ -0,0 +1,723 @@
+    vf_coef: float = 0.1
+    """the value function coefficient"""
+    cliprange: float = 0.2
+    """the clip range"""
+    cliprange_value: float = 0.2
+    """the clip range for the value function"""
+    gamma: float = 1
+    """the discount factor"""
+    lam: float = 0.95
+    """the lambda value for GAE"""
+    whiten_rewards: bool = False
+    """whether to whiten the rewards"""
+    kl_coef: float = 0.05
+    """the KL coefficient"""
+
+    # rloo args
+    rloo_k: int = 4
+    """REINFORCE Leave-One-Out (RLOO) number of online samples per prompt"""
+
+
+# taken from https://github.com/vwxyzjn/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/utils.py#L99
+def disable_dropout(model: torch.nn.Module) -> None:
+    """Disable dropout in a model."""
+    for module in model.modules():
+        if isinstance(module, torch.nn.Dropout):
+            module.p = 0
+
+
+def get_reward(model, query_responses, tokenizer, context_length):
+    attention_mask = query_responses != tokenizer.pad_token_id
+    # position_ids = attention_mask.cumsum(1) - attention_mask.long()  # exclusive cumsum
+    lm_backbone = getattr(model, model.base_model_prefix)
+    input_ids = torch.masked_fill(query_responses, ~attention_mask, 0)
+    output = lm_backbone(
+        input_ids=input_ids,
+        attention_mask=attention_mask,
+        # position_ids=position_ids,
+        return_dict=True,
+        output_hidden_states=True,
+    )
+    reward_logits = model.score(output.hidden_states[-1])
+    sequence_lengths = (
+        first_true_indices(query_responses[:, context_length:] == tokenizer.pad_token_id) - 1 + context_length
+    )
+    # https://github.com/huggingface/transformers/blob/dc68a39c8111217683bf49a4912d0c9018bab33d/src/transformers/models/gpt2/modeling_gpt2.py#L1454
+    return (
+        reward_logits,
+        reward_logits[
+            torch.arange(reward_logits.size(0), device=reward_logits.device),
+            sequence_lengths,
+        ].squeeze(-1),
+        sequence_lengths,
+    )
+
+
+def exact_div(a, b):
+    q = a // b
+    if a != q * b:
+        raise ValueError(f"Inexact division: {a} / {b} = {a / b}")
+    return q
+
+
+def generate(lm_backbone, queries, tokenizer, generation_config):
+    """generate in a way that does not affect padding tokens"""
+    context_length = queries.shape[1]
+    attention_mask = queries != tokenizer.pad_token_id
+    input_ids = torch.masked_fill(queries, ~attention_mask, 0)
+    output = lm_backbone.generate(
+        input_ids=input_ids,
+        attention_mask=attention_mask,
+        # position_ids=attention_mask.cumsum(1) - attention_mask.long(), # not needed: already adjusted in generations
+        # https://github.com/huggingface/transformers/blob/ac33aeeeee2a7a89b89c93c2962e6feb90daef0a/src/transformers/models/gpt2/modeling_gpt2.py#L1227-L1250
+        generation_config=generation_config,
+        return_dict_in_generate=True,
+        output_scores=True,
```

Comment:
Why do we use `output_scores` instead of `output_logits`?

Reply:
Good point! We can know (since this is a new API :) )

Reply:
As I understand it the difference between scores and logits is the application of logitprocessor.

Sampling with temp 1.0, scores != logits
```
>>> inputs = tokenizer("Hello, please tell me if the logits work", return_tensors="pt")
>>> output = base_model.generate(**inputs, output_logits=True, output_scores=True, return_dict_in_generate=True, do_sample=True, temperature=1.0)
>>> >>> torch.max(F.softmax(torch.stack(output.logits), dim=2), dim=2).values
tensor([[0.1969],
        [0.2204],
        [0.2907],
        [0.3935],
        [0.8413],
        [0.9996],
        [0.9149],
        [0.1471],
        [0.3852]])
>>> torch.max(F.softmax(torch.stack(output.scores), dim=2), dim=2).values
tensor([[0.2024],
        [0.2977],
        [0.3351],
        [0.4187],
        [0.8420],
        [0.9997],
        [0.9149],
        [0.1526],
        [0.4085]])
```

No sampling, logits and scores are equal
```
>>> output = base_model.generate(**inputs, output_logits=True, output_scores=True, return_dict_in_generate=True, do_sample=False)
>>> torch.max(F.softmax(torch.stack(output.logits), dim=2), dim=2).values
tensor([[0.1969],
        [0.4525],
        [0.3918],
        [0.9702],
        [0.7385],
        [0.6670],
        [0.1989],
        [0.9851],
        [0.3038]])
>>> torch.max(F.softmax(torch.stack(output.scores), dim=2), dim=2).values
tensor([[0.1969],
        [0.4525],
        [0.3918],
        [0.9702],
        [0.7385],
        [0.6670],
        [0.1989],
        [0.9851],
        [0.3038]])
````

I noticed this when I saw a difference between the `generate()` logits and the `forward()` logits.